### ISSUE TITLE
Security: can't secure API

### ISSUE THREAD

**Issue by pascalgrimaud** [jhipster/generator-jhipster#7374]
##### **Overview of the issue**

Currently, in master branch, we can't secure an API with:

```java
@Secured(AuthoritiesConstants.ADMIN)
```

I will complete this ticket later, if needed

##### **Reproduce the error**

- create an entity
- secure all API, using `@Secured(AuthoritiesConstants.ADMIN)`
- connect with user/user
- you can access to the entity

##### **Browsers and Operating System**

- [ ] Checking this box is mandatory (this is just to show you read everything)

**Comment by deepu105** [comment]
@pascalgrimaud does `@PreAuthorize("hasRole('ROLE_USER')")` work?

**Comment by pascalgrimaud** [comment]
I will test.
I created the ticket during the formation, with Julien and Pierre, didn't investigate enough.

**Comment by jdubois** [comment]
I've found the cause, this is crashing because of:
```
@Bean
    @Override
    public AuthenticationManager authenticationManager() {
```

In the `SecurityConfiguration` Bean. This was working with JHipster 4, however.

**Comment by jdubois** [comment]
So this "AuthenticationManager" part used to be configured using a `@PostConstruct` annotation and now this is a Spring Bean, which makes this fail.

The smartass who did this change is me, of course :-)
And I have no idea why :-))

The issue is that this change looks required by Spring Boot 2.0, and if we don't do it, it looks like we have no AuthenticationManager anymore. So there seems to be a good reason initially to do this change.

**Comment by pascalgrimaud** [comment]
Thanks Julien for the full details.

As discussed yesterday, I confirm we have the same issue in Registry, but as there is only 1 ROLE_ADMIN there, it's not really important.

Here the 2 links of Security part with Spring Boot 2, in the case of someone wants to help:
https://github.com/spring-projects/spring-boot/wiki/Spring-Boot-2.0-Migration-Guide#security
https://github.com/spring-projects/spring-boot/wiki/Spring-Boot-Security-2.0

**Comment by pascalgrimaud** [comment]
@deepu105 : just tried your suggestion, it doesn't work, because `@EnableGlobalMethodSecurity(prePostEnabled = true, securedEnabled = true)` is missing

**Comment by juliensadaoui** [comment]
I found a fix for this problem. The AuthenticationManager bean is injected into UserJWTController bean. With JHipster 4, this "AuthenticationManager" is configured using the `@PostContruct` annotation. I think that it's not a good idea to create directly the AuthenticationManager. I prefer to delegate this instantiation using a AuthenticationManagerBean, the bean is created by Spring Security. See the documentation for more details. 
https://github.com/spring-projects/spring-boot/wiki/Spring-Boot-2.0-Migration-Guide#authenticationmanager-bean

```java
@Override
@Bean
public AuthenticationManager authenticationManagerBean() throws Exception {
        return super.authenticationManagerBean();
}
```

**Comment by RothAndrew** [comment]
Any update on this? I'm still seeing this issue with v5.2.0. Using the `@Secured` annotation works fine "In Real Life", but none of the tests that JHipster generates will fail if I add the `@Secured` annotation and use `@WithMockUser` in the test.

Steps to reproduce:
* Generate an app (I'm currently working with a microservice gateway using JHipster 5.2.0)
* Add `@Secured(AuthoritiesConstants.ADMIN)` to a rest method that didn't already have it.
* Add `@WithMockUser(username = "user", password = "password", roles = "USER") to a corresponding test, so that it __should__ fail (but doesn't)

**Comment by jdubois** [comment]
For tests it's a different issue, it's just that the Spring test context doesn't configure security. I guess the idea is that we just test business methods, and also that'll easier to setup and faster to run.
But I'm like you, I would like to be able to test this, for me that's a limitation with Spring tests.

**Comment by mraible** [comment]
In my [JHipster 5 tutorial](https://github.com/mraible/jhipster5-demo), I was able to use [`@WithMockUser`](https://github.com/mraible/jhipster5-demo#deploy-to-the-cloud) w/o any issues. I'm not using the `@Secured` annotation, but `@WithMockUser` works for me.

**Comment by jdubois** [comment]
Oh yes, @WithMockUser works, I use it very often. It's only the @Secured.

**Comment by jdubois** [comment]
Of course @cbornet has the best answer: https://github.com/jhipster/generator-jhipster/issues/7806#issuecomment-398165015

**Comment by RothAndrew** [comment]
I got it to work, without having to use @cbornet's suggestion in https://github.com/jhipster/generator-jhipster/issues/7806#issuecomment-398165015, which was throwing an exception (see https://github.com/jhipster/generator-jhipster/issues/7806#issuecomment-417109949).

I annotated the service method with `@Secured(AuthoritiesConstants.ADMIN)`, rather than the rest controller method. It actually feels better to do it that way anyway, so I really don't have to change the rest controllers at all and all of my logic lives in the service.

Even better than that, I was able to annotate the service __interface's__ method, rather than the service method, and it worked.
